Sources/BuildServerIntegration/BuildServerManager.swift

```diff
@@ -962,6 +962,123 @@ package actor BuildServerManager: QueueBasedMessageHandler {
     return locations.map { locationAdjustedForCopiedFiles($0) }
   }
 
+  private func uriAdjustedForCopiedFiles(_ uri: DocumentURI) -> DocumentURI {
+    guard let originalUri = cachedCopiedFileMap[uri] else {
+      return uri
+    }
+    return originalUri
+  }
+
+  package func workspaceEditAdjustedForCopiedFiles(_ workspaceEdit: WorkspaceEdit?) -> WorkspaceEdit? {
+    guard var edit = workspaceEdit else {
+      return nil
+    }
+    if let changes = edit.changes {
+      var newChanges: [DocumentURI: [TextEdit]] = [:]
+      for (uri, edits) in changes {
+        let newUri = self.uriAdjustedForCopiedFiles(uri)
+        newChanges[newUri, default: []] += edits
+      }
+      edit.changes = newChanges
+    }
+    if let documentChanges = edit.documentChanges {
+      var newDocumentChanges: [WorkspaceEditDocumentChange] = []
+      for change in documentChanges {
```

Comment:
Nitpick: Could this use `map` or `reserverCapacity` since we already know how many elements we’ll need in `newDocumentChanges`? Same for `newChanges` and in `locationsOrLocationLinksAdjustedForCopiedFiles`.